pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,239 @@ func (q *QueryGVR) Name() string {
 }
 
 // WithVersions checks if an API group with all the specified versions exist.
+// This method can be omitted to query any version.
 func (q *QueryGVR) WithVersions(versions ...string) *QueryGVR {
 	q.versions = versions
 	return q
 }
 
 // WithResource checks if an API group with the specified resource exists.
-// WithVersions needs to be used before calling this function.
+// This method can be omitted to query any resource.
 func (q *QueryGVR) WithResource(resource string) *QueryGVR {
-	q.resource = resource
+	q.resource.set(resource)
 	return q
 }
 
-// Run discovery
+// Run discovery.
 func (q *QueryGVR) Run(config *clusterQueryClientConfig) (bool, error) {
-	if q.group != "" && q.resource != "" && len(q.versions) == 0 {
-		return false, fmt.Errorf("cannot check for group and resource existence without version info; use WithVersion method")
+	if err := q.validate(config); err != nil {
+		return false, fmt.Errorf("validating query: %w", err)
```

Comment:
nit: error message needs to add sufficient context when you are wrapping. You could change this to `query validation failed: %w` (still not additional context IMO) or remove the error wrapping entirely.

Reply:
I'm against putting the word "failed" or "error" in an error message. It is redundant as it is clear in both the output and in the code that the message is due to an error. The Thanos Style Guide [expresses this sentiment](https://github.com/thanos-io/thanos/blob/main/docs/contributing/coding-style-guide.md#wrap-errors-for-more-context-dont-repeat-failed--there) more strongly than I do.

However, I agree more context can be added. I'll change it to something like,

```go
fmt.Errorf("GroupVersionResource API query validation: %w", err)
```

Reply:
Not sure I agree with the Thanos style guide when multiple errors are wrapped, but this looks good:

```
fmt.Errorf("GroupVersionResource API query validation: %w", err)
```

Reply:
Error message has been updated with more context.